Improve "Submit Feedback" Flow

kean opened the pull request:
## Changes

- Add "Submit Feedback" button to the "Me" menu so that users can submit feedback at any time
- Add an optional "Subject" field to the form. Currently all requests have the default subject "Jetpack for iOS Support" A custom subject will make it easier to navigate these. The limit is 100 characters.
- Add 500 character limit to the "Details" field – this should be enough and limiting it will reduce the time needed to go through the feedbacks.
- Add support for adding basic image and GIF attachments. The limit 5 attachments, the maximum size is 8 MB. The app uses the existing `MediaImageExporter` infra to downscale the attachments to 1024×1024 pixels and remove location data.
- Stop asking to confirm your email and name when submitting feedbacks – there are supposed to be one-way interactions
- Add a confirmation dialog on "Cancel" and disable interactive dismiss when non-empty
- Add haptics on submit success and failure
- (Unrelated) Update the "Share Jetpack" Me menu icon to match the style of other icons

Example feedback with attachments created using the new form: 8465509-zd-a8c.

## Screenshots

**Notes on design**

- Using `.plain` list style to make sure there is enough space on smaller devices to fit the form. This is a common design for these types of screens on iOS, so nobody should be confused.
- `TextEditor` doesn't support placeholders, so it'll have to be like this – should be fine
- On iPad, it is presented in a [sheet](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/0775d305-0615-4416-b2fc-da62c59acfca)

<img width="240" alt="Screenshot 2024-07-10 at 2 10 07 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/d0260642-dcf1-41b4-af2e-107b7317ef04">
<img width="240" alt="Screenshot 2024-07-10 at 1 15 47 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/c4fefbdf-c13f-48e2-a236-49d4809f8c61">
<img width="240" alt="Screenshot 2024-07-10 at 1 33 28 PM" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/066fcf08-78e4-43d4-acc5-d086224451ae">

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/bd5cd6df-3284-4816-9e45-26bf6933f2d8

## To test:

- Verify that you can submit a feedback with and without the attachments
- Check the acceptable criteria based on the "Changes" list

**No Prior Support Interactions**

- Login in the app for the first time
- Tap "Submit Feedback"
- Verify that you can upload attachments
- Submit a feedback
- Verify that the feedback was uploaded and reached Zendesk
- Verify that no prompts to confirm your email and/or name were shown
- Open "Help & Support" / "Tickets" and verify that now it asks to confirm your email and name

## Regression Notes
1. Potential unintended areas of impact: "Submit Feedback"
2. What I did to test those areas of impact (or what existing automated tests I relied on): n/a
3. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

kean committed:
Add subject support to SubmitFeedbackViewController

kean committed:
Add support for attachments

kean committed:
Add Submit Feedback button to the Me menu

kean committed:
Remove redundant tag

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

kean committed:
Show as a sheet on iPad and reduce maxmimum attachment size

jkmassel reviewed (changes requested):
If I haven't previously submitted feedback, I get a popup after filling out the form asking me for my name and email address – can we put these on the form as optional fields? Relatedly, if I don't have a feedback identity set, I'm not able to upload images (it just says "forbidden" the first time – it works on subsequent runs).

edit: If they're logged in with WordPress.com, I'd imagine we could just use those credentials? And if they're not, the feedback could be anonymous? WDYT?

kean commented:
> Relatedly, if I don't have a feedback identity set, I'm not able to upload images (it just says "forbidden" the first time – it works on subsequent runs).

Oh, good catch. I tested with identity already set.

> If they're logged in with WordPress.com, I'd imagine we could just use those credentials? And if they're not, the feedback could be anonymous? WDYT?

I'll look into it. It was an existing part of the flow, so I haven't touched it.

kean committed:
Create identifity silently for Zendesk feedback request

kean commented:
> edit: If they're logged in with WordPress.com, I'd imagine we could just use those credentials? And if they're not, the feedback could be anonymous? WDYT?

I implemented it in https://github.com/wordpress-mobile/WordPress-iOS/pull/23414/commits/36e2780d7ae23b02df8bae681e6eacca8630a5c2 and added the testing instruction to the PR. It's a nice improvement because it's one less step. I couldn't think of any reasons to confirm the user email and name at this point as these interactions are supposed to be one-way – we are not setting any expectations about responding to the feedback.

I haven't done enough testing yet, but the happy path for the wp.com accounts work well.

kean commented on WordPress/Classes/Utility/ZendeskUtils.swift:
```diff
@@ -418,6 +418,28 @@ extension ZendeskUtils {
+    /// such as sending feedback about the app.
+    @MainActor func createIdentifitySilently() async {
+        // If we already have an identity, and the user has confirmed it, do nothing.
+        let haveUserInfo = ZendeskUtils.sharedInstance.haveUserIdentity && ZendeskUtils.sharedInstance.userNameConfirmed
```
It's a bit of duplication, but I'm not up to refactoring this whole class. The priority was to not break any of the existing flows.

kean commented:
Moving back to draft as there is one more change needed to update how it works with identities.

kean commented:
I updated the [previous implementation](https://github.com/wordpress-mobile/WordPress-iOS/pull/23414#issuecomment-2221764071). I tried to mess with `ZendeskUtils` as little as possible because it's fairly complicated, and I didn't want to spend too much time on it. I tested the following, so I know at least that's true:

- The app calls [`ZendeskUtils.fetchUserInformation()`](https://github.com/wordpress-mobile/WordPress-iOS/blob/2a20648fe23bdc50c279c793a166b2b883597a8f/WordPress/Classes/Utility/ZendeskUtils.swift#L111) on app launch and it _usually_ creates an identity before you reach "Submit Feedback".
- `ZendeskUtils.fetchUserInformation` is not called after you login.
- There are some parts of `ZendeskUtils.fetchUserInformation` that are asynchronous, and [some parts](https://github.com/wordpress-mobile/WordPress-iOS/blob/2a20648fe23bdc50c279c793a166b2b883597a8f/WordPress/Classes/Utility/ZendeskUtils.swift#L589) where the `completion` and the "create identity" logic might end up never being called. There is no guarantee that `Zendesk.instance.identity` will be set.
- `Zendesk.instance.identity` _seems_ to be the only thing required to submit a feedback and to upload attachments. `Identity.createAnonymous()` will also do. Here's an example feedback I submitted using an anonymous session: 8469179-zd-a8c.

Given all that, I added some defensive code to the form that will ensure that if no `Zendesk.instance.identity` is set, it create an anonymous one. In most cases, the identity _will_ be set, so it will be no-op.

I tested the following scenarios:

- Fresh install, add self-hosted site, go straight to "Submit Feedback" and submit feedback
- Fresh install, add self-hosted site, add wp.com account, and submit feedback
- Fresh install, login with a wp.com account, and submit feedback

I also found a bug with a GDPR popover during testing self-hosted sites https://github.com/wordpress-mobile/WordPress-iOS/issues/23416.

jkmassel reviewed (approved):
I tested this on-device and it works well.

To change my Zendesk identity I had to delete and re-install the app, and I don't love that we might not be able to reply to someone's feedback. I think it's out of the scope of this PR though – we'll solve that in a subsequent one.
